File: python/synor/ops/code.py
r"""Structural code matching over a reusable parsed AST.

Wrap source in a :class:`CodeSource`, then pass it to any parse-consuming API —
structural pattern matching (:class:`CodePattern`, :func:`match_code`),
chunk splitting (:meth:`RecursiveSplitter.split
<synor.ops.text.RecursiveSplitter.split>`), prefilter-index term extraction
(:func:`index_terms`) — and the source is parsed **at most once**.
Metavariables in a pattern use the ``\`` sigil (e.g. ``\NAME``, ``\(ARGS*\)``).
"""

__all__ = [
    "CodeMatch",
    "CodePattern",
    "CodeSource",
    "FileMatch",
    "index_terms",
    "match_code",
    "render_match",
]

import typing as _typing
import warnings as _warnings
from dataclasses import dataclass as _dataclass

from synor._internal import core as _core
from synor.resources import chunk as _chunk
from synor.resources import view as _view


class CodeSource:
    """Source text plus a lazily-parsed, shared AST.

    The universal input for APIs that may need a parse: pass one handle to
    several of them (:meth:`RecursiveSplitter.split
    <synor.ops.text.RecursiveSplitter.split>`,
    :meth:`CodePattern.match_source`, :func:`match_code`,
    :func:`index_terms`, …) and the source is parsed **at most once**, no
    matter how many consumers touch it.

    Construction never parses and never raises: an unknown or non-tree-sitter
    language is fine — each consumer takes its own degraded (non-AST) path,
    exactly as if it had been given a plain ``str``, or raises at call time if
    it genuinely requires an AST.

    Args:
        text: The source text.
        language: Language name, alias, or file extension (e.g. ``"python"``,
            ``"c++"``, ``".rs"``). ``None`` means no syntax awareness.

    Examples:
        >>> src = CodeSource("def f(): pass", language="python")
        >>> splitter = RecursiveSplitter()          # doctest: +SKIP
        >>> chunks = splitter.split(src, chunk_size=1000)   # parses once
        >>> cp = CodePattern(r"def \\NAME(\\(A*\\)):", language="python")
        >>> ms = cp.match_source(src)               # reuses the same parse
    """

    def __init__(self, text: str, language: str | None = None) -> None:
        self._src = _core.CodeSource(text, language)

    @classmethod
    def _from_core(cls, raw: "_core.CodeSource") -> "CodeSource":
        """Wrap an already-built core source (e.g. from ``CodePattern.match_file``,
        with its parse already cached) without copying."""
        self = cls.__new__(cls)
        self._src = raw
        return self

    @property
    def text(self) -> str:
        """The source text."""
        return self._src.text

    @property
    def language(self) -> str | None:
        """The language as given at construction (may be an alias/extension)."""
        return self._src.language

    @property
    def source(self) -> str:
        """Deprecated alias of :attr:`text` (the name the removed ``CodeAst``
        class used); use ``.text``."""
        _warnings.warn(
            "CodeSource.source is deprecated; use CodeSource.text",
            DeprecationWarning,
            stacklevel=2,
        )
        return self._src.text

    def __repr__(self) -> str:
        return f"CodeSource(language={self.language!r}, text_len={len(self.text)})"


@_dataclass(frozen=True, slots=True)
class CodeMatch:
    """A structural-match result: a matched node and its captured metavariables."""

    kind: str
    """tree-sitter node kind of the matched node (e.g. ``function_definition``)."""

    chunks: list[_chunk.Chunk]
    """The matched code region(s), each with text and line/column positions.

    Currently always exactly one chunk (the whole matched node); a future
    carve-out feature may return several (e.g. a function's head and tail with the
    body elided)."""

    captures: dict[str, list[_chunk.Chunk]]
    """Captured metavariables: name -> matched region(s).

    Each captured value is a list of chunks (currently always exactly one); the
    captured text is ``m.captures[name][0].text``, with line/column on each chunk."""


class CodePattern:
    r"""A compiled structural pattern, built once and reused across many sources.

    Compiling a pattern (and its prefilter) is not free; construct a ``CodePattern``
    once and match it against many files/sources instead of calling
    :func:`match_code` with a pattern string every time (which recompiles).

    Args:
        pattern: The by-example structural pattern (``\`` sigil for metavars).
        language: Language name or alias (e.g. ``"python"``, ``"c++"``).
        min_len: Prefilter tuning — required terms shorter than this are dropped.

    Examples:
        >>> cp = CodePattern(r"def \NAME(\(A*\)):", language="python")
        >>> cp.might_match("x = 1")          # cheap, parse-free pre-check
        False
        >>> [m.captures["NAME"][0].text for m in cp.match_source("def f(): pass")]
        ['f']
    """

    def __init__(self, pattern: str, language: str, *, min_len: int = 3) -> None:
        self._cp = _core.CodePattern(pattern, language, min_len)

    @property
    def language(self) -> str:
        """The language this pattern was compiled for."""
        return self._cp.language

    def might_match(self, source: str) -> bool:
        """Whether ``source`` *might* contain a match — a cheap, parse-free prefilter
        check. ``False`` means it definitely can't (skip it); ``True`` means "maybe"."""
        return self._cp.might_match(source)

    def match_source(self, source: "str | CodeSource") -> list[CodeMatch]:
        """Match against ``source`` — a ``str`` (parsed on the spot) or a
        :class:`CodeSource` (reusing its cached parse) — skipping the parse
        entirely when the prefilter rejects it. Reuses this pattern's
        compilation across calls."""
        if isinstance(source, CodeSource):
            raw, text = self._cp.match_source(source._src), source.text
        else:
            raw, text = self._cp.match_source(source), source
        return [_convert_match(m, text) for m in raw]

    def match_file(self, path: str) -> "FileMatch | None":
        """Read ``path``, prefilter, and (only if it might match) parse + match.

        Returns a :class:`FileMatch` (parsed source + matches) when there is at
        least one match, else ``None`` — so a rejected or non-matching file never
        costs a parse beyond what the prefilter needs. Non-UTF-8 (binary) files are
        skipped (``None``); other I/O errors raise ``OSError``.
        """
        raw = self._cp.match_file(path)
        if raw is None:
            return None
        source = CodeSource._from_core(raw.source)
        return FileMatch(
            path=raw.path,
            source=source,
            matches=[_convert_match(m, source.text) for m in raw.matches],
        )


@_dataclass(frozen=True, slots=True)
class FileMatch:
    """The result of :meth:`CodePattern.match_file`: the parsed source and the
    matches found in one file. The file content is ``file_match.source.text``."""

    path: str
    """The path that was matched."""

    source: CodeSource
    """The parsed source (its AST is already cached — reuse it to split or match
    more patterns without re-parsing)."""

    matches: list[CodeMatch]
    """The matches found (always at least one — ``match_file`` returns ``None``
    when there are none)."""

    @property
    def ast(self) -> CodeSource:
        """Deprecated alias of :attr:`source` (the field's name when it held the
        removed ``CodeAst`` class); use ``.source``."""
        _warnings.warn(
            "FileMatch.ast is deprecated; use FileMatch.source "
            "(and .text for the file content)",
            DeprecationWarning,
            stacklevel=2,
        )
        return self.source


def match_code(
    pattern: str, source: "str | CodeSource", language: str | None = None
) -> list[CodeMatch]:
    r"""One-shot: match ``pattern`` against ``source`` and return all matches.

    ``source`` is a ``str`` (with ``language`` required) or a :class:`CodeSource`
    (whose cached parse is reused; ``language`` must be omitted). Prefer a
    :class:`CodePattern` when matching the same pattern across many sources —
    this recompiles the pattern per call."""
    if isinstance(source, CodeSource):
        raw, text = _core.match_code(pattern, source._src, language), source.text
    else:
        raw, text = _core.match_code(pattern, source, language), source
    return [_convert_match(m, text) for m in raw]


def index_terms(
    source: "str | CodeSource", language: str | None = None, min_len: int = 3
) -> list[str]:
    """Extract the indexable terms of ``source`` (identifiers + string-literal
    content, ``>= min_len`` chars, deduped), for building an external prefilter
    index (FTS / n-grams).

    ``source`` is a ``str`` (with ``language`` required) or a :class:`CodeSource`
    (whose cached parse is reused; ``language`` must be omitted). Raises
    ``ValueError`` for an unknown or non-tree-sitter language — silently
    indexing nothing would poison the index with false negatives."""
    raw = source._src if isinstance(source, CodeSource) else source
    return _core.index_terms(raw, language, min_len)


def _convert_match(raw: "_core.CodeMatch", source: str) -> CodeMatch:
    """Convert a raw PyO3 match to a Python dataclass."""
    return CodeMatch(
        kind=raw.kind,
        chunks=[_convert_chunk(c, source) for c in raw.chunks],
        captures={
            name: [_convert_chunk(c, source) for c in chunks]
            for name, chunks in raw.captures.items()
        },
    )


def _convert_chunk(raw: "_core.Chunk", text: str) -> _chunk.Chunk:
    """Convert a raw PyO3 chunk to a Python Chunk dataclass."""
    chunk_text = text[raw.start_char_offset : raw.end_char_offset]
    return _chunk.Chunk(
        text=chunk_text,
        start=_chunk.TextPosition(
            byte_offset=raw.start_byte,
            char_offset=raw.start_char_offset,
            line=raw.start_line,
            column=raw.start_column,
        ),
        end=_chunk.TextPosition(
            byte_offset=raw.end_byte,
            char_offset=raw.end_char_offset,
            line=raw.end_line,
            column=raw.end_column,
        ),
    )


def render_match(source: "CodeSource", match: CodeMatch) -> _view.SourceView:
    r"""Render a match as a source view: context frames of its enclosing scopes,
    the matched range(s) verbatim, and elision cues where material is omitted.

    The returned :class:`~synor.resources.view.SourceView` carries synthetic
    ``text`` plus source-grounded segments — frames are the head lines of
    enclosing scopes, outermost first, and its ``start``/``end`` (frames
    excluded) equal the match's span. Matched ranges are rendered exactly (no
    widening to line starts).

    Args:
        source: The parsed source the match came from (e.g.
            ``file_match.source``); its cached parse is reused.
        match: A match produced against that same source.

    Examples:
        >>> fm = CodePattern(r"return \X", language="python").match_file(path)
        >>> view = render_match(fm.source, fm.matches[0])
        >>> print(view.text)  # frames + matched code
    """
    ranges = [(c.start.byte_offset, c.end.byte_offset) for c in match.chunks]
    raw = _core.render_ranges(source._src, ranges)
    return _convert_source_view(raw)


def _convert_source_view(raw: "_core.SourceView") -> _view.SourceView:
    """Convert a raw PyO3 source view to a Python dataclass."""
    segments = [
        _view.ViewSegment(
            start=_chunk.TextPosition(
                byte_offset=seg.start_byte,
                char_offset=seg.start_char_offset,
                line=seg.start_line,
                column=seg.start_column,
            ),
            end=_chunk.TextPosition(
                byte_offset=seg.end_byte,
                char_offset=seg.end_char_offset,
                line=seg.end_line,
                column=seg.end_column,
            ),
            kind=_typing.cast('_typing.Literal["frame", "content"]', seg.kind),
            summary=seg.summary,
            rendered_start=seg.rendered_start,
            rendered_end=seg.rendered_end,
        )
        for seg in raw.segments
    ]
    return _view.SourceView(text=raw.text, segments=segments)


def __getattr__(name: str) -> _typing.Any:
    # Deprecated alias for the removed ``CodeAst`` class, so annotations and
    # ``isinstance`` checks in older callers keep working. ``CodeSource`` is
    # the single handle now; note its constructor is lazy and tolerant where
    # ``CodeAst``'s was eager and raising, and the old ``.matches`` / ``.split``
    # / ``.index_terms`` methods live on the consumers instead
    # (``CodePattern.match_source`` / ``match_code``, ``RecursiveSplitter.split``,
    # ``index_terms``).
    if name == "CodeAst":
        _warnings.warn(
            "CodeAst is deprecated and now aliases CodeSource; use CodeSource "
            "with CodePattern.match_source / match_code, RecursiveSplitter.split, "
            "and index_terms",
            DeprecationWarning,
            stacklevel=2,
        )
        return CodeSource
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")

File: rust/sdk/synor/src/ops/code.rs
//! Structural code matching over a reusable parsed AST.
//!
//! Mirrors Python's `synor.ops.code`. Parse source once into a [`CodeAst`],
//! then match by-example structural patterns and/or split without re-parsing.
//! Metavariables in a pattern use the `\` sigil (e.g. `\NAME`, `\(ARGS*\)`).
//!
//! Enabled by the `code_match` feature.
//!
//! # Examples
//!
//! ```rust
//! # #[cfg(feature = "code_match")] {
//! use synor::ops::code::{CodeAst, CodePattern, match_code};
//!
//! // One-shot: parse + match in one call.
//! let matches = match_code(r"\NAME(\(A*\))", "def f(x): pass", "python").unwrap();
//! assert!(!matches.is_empty());
//! assert_eq!(matches[0].captures["NAME"][0].text("def f(x): pass"), "f");
//!
//! // Reuse the parse across multiple patterns.
//! let ast = CodeAst::new("def f(x): return x", "python").unwrap();
//! let m = ast.matches(r"def \NAME(\(A*\)):").unwrap();
//! assert_eq!(m[0].captures["NAME"][0].text(ast.source()), "f");
//!
//! // Reuse the compiled pattern across many sources.
//! let pat = CodePattern::new(r"def \NAME(\(A*\)):", "python", 3).unwrap();
//! assert!(pat.might_match("def foo(): pass"));
//! assert!(!pat.might_match("x = 1"));
//! # }
//! ```

use std::collections::HashMap;

use synor_code_ast::tree_sitter::Tree;
use synor_code_ast::{LineIndex, OutputPosition, ParseOutcome};
use synor_code_match::lang;
use synor_code_match::{Pattern, Prefilter, index_terms_in_tree};
use synor_ops_text::split::{RecursiveChunkConfig, RecursiveChunker, RecursiveSplitConfig};

use crate::error::{Error, Result};
use crate::resources::chunk::{Chunk, TextPosition};

pub use synor_code_ast::CodeSource;

// ─── Internal helpers ────────────────────────────────────────────────────────

fn pos_from(byte_offset: usize, op: OutputPosition) -> TextPosition {
    TextPosition {
        byte_offset,
        char_offset: op.char_offset,
        line: op.line,
        column: op.column,
    }
}

fn build_chunk(line_index: &LineIndex, source: &str, start_byte: usize, end_byte: usize) -> Chunk {
    let sp = line_index.position(source, start_byte);
    let ep = line_index.position(source, end_byte);
    Chunk::new(
        start_byte..end_byte,
        pos_from(start_byte, sp),
        pos_from(end_byte, ep),
    )
}

fn build_sdk_matches(
    source: &str,
    line_index: &LineIndex,
    raw: Vec<synor_code_match::Match<'_>>,
) -> Vec<CodeMatch> {
    raw.into_iter()
        .map(|m| {
            let chunk = build_chunk(line_index, source, m.range.start, m.range.end);
            let captures = m
                .captures
                .iter()
                .map(|(name, c)| {
                    let ch = build_chunk(line_index, source, c.range.start, c.range.end);
                    (name.clone(), vec![ch])
                })
                .collect();
            CodeMatch {
                kind: m.kind,
                chunks: vec![chunk],
                captures,
            }
        })
        .collect()
}

// ─── Public types ────────────────────────────────────────────────────────────

/// A structural-match result: the matched node and its captured metavariables.
///
/// Chunks are range-only (no owned text). Call `chunk.text(source)` with the
/// source string from [`CodeAst::source`] or your own variable to obtain the
/// slice.
pub struct CodeMatch {
    /// tree-sitter node kind of the matched node (e.g. `"function_definition"`).
    pub kind: &'static str,
    /// The matched code region(s). Currently always exactly one (the whole matched
    /// node); future versions may produce several (e.g. head + tail of a function
    /// with the body elided).
    pub chunks: Vec<Chunk>,
    /// Captured metavariables: name → matched region(s). Each value is currently a
    /// single-element list; the list shape leaves room for future multi-region
    /// captures. Get the text with `m.captures["NAME"][0].text(source)`.
    pub captures: HashMap<String, Vec<Chunk>>,
}

/// A compiled structural pattern + prefilter, built once and reused across many
/// sources or [`CodeAst`]s.
///
/// Compiling a pattern is not free: construct a `CodePattern` once and call
/// [`match_source`](CodePattern::match_source) or
/// [`match_file`](CodePattern::match_file) many times rather than reparsing with
/// [`match_code`] on each source.
pub struct CodePattern {
    language: String,
    pattern: Pattern,
    prefilter: Prefilter,
}

impl CodePattern {
    /// Compile `pattern` for `language` with a custom `min_len` prefilter tuning.
    ///
    /// `min_len` controls the minimum term length considered by the prefilter;
    /// smaller values increase recall (fewer false rejections) at the cost of speed.
    /// The default (used by [`CodePattern::compile`]) is 3.
    pub fn new(pattern: &str, language: impl Into<String>, min_len: usize) -> Result<Self> {
        let language = language.into();
        let cfg = lang::by_name(&language).ok_or_else(|| {
            Error::engine(format!(
                "structural matching is not supported for language {language:?}"
            ))
        })?;
        let compiled = Pattern::compile(pattern, &cfg)?;
        let prefilter = compiled.prefilter(min_len);
        Ok(Self {
            language,
            pattern: compiled,
            prefilter,
        })
    }

    /// Compile with the default `min_len` of 3.
    pub fn compile(pattern: &str, language: impl Into<String>) -> Result<Self> {
        Self::new(pattern, language, 3)
    }

    /// The language this pattern was compiled for.
    pub fn language(&self) -> &str {
        &self.language
    }

    /// Whether `source` *might* contain a match — a cheap, parse-free prefilter
    /// check. `false` means it definitely cannot match (skip the source entirely);
    /// `true` means "maybe" and a full parse is needed to confirm.
    pub fn might_match(&self, source: &str) -> bool {
        self.prefilter.might_match(source)
    }

    /// Match against `source`, parsing it fresh. Skips the parse entirely when the
    /// prefilter rejects `source`. Reuses this pattern's compilation across calls.
    pub fn match_source(&self, source: &str) -> Vec<CodeMatch> {
        let raw = self.pattern.matches_prefiltered(source, &self.prefilter);
        if raw.is_empty() {
            return Vec::new();
        }
        let line_index = LineIndex::build(source);
        build_sdk_matches(source, &line_index, raw)
    }

    /// Read the file at `path`, run the prefilter (parse-free), and — only if it
    /// might match — parse and match.
    ///
    /// Returns `Some(`[`FileMatch`]`)` with the parsed AST and matches when there
    /// is at least one match, `None` when the file is rejected or has no matches.
    /// Non-UTF-8 (binary) files are silently skipped (`None`); other I/O errors
    /// propagate as `Err`.
    pub fn match_file(&self, path: &str) -> Result<Option<FileMatch>> {
        let content = match std::fs::read_to_string(path) {
            Ok(c) => c,
            Err(e) if e.kind() == std::io::ErrorKind::InvalidData => return Ok(None),
            Err(e) => {
                return Err(Error::engine(format!("failed to read {path:?}: {e}")));
            }
        };
        if !self.prefilter.might_match(&content) {
            return Ok(None);
        }
        let source = CodeSource::with_language(content, self.language.clone());
        let raw = self.pattern.matches_source(&source);
        if raw.is_empty() {
            return Ok(None);
        }
        let matches = build_sdk_matches(source.text(), source.line_index(), raw);
        Ok(Some(FileMatch {
            path: path.to_string(),
            ast: CodeAst { source },
            matches,
        }))
    }
}

/// A parsed code AST: parse once, then [`matches`](CodeAst::matches),
/// [`matches_with`](CodeAst::matches_with), and/or [`split`](CodeAst::split)
/// without re-parsing. A thin, eagerly-validated wrapper over [`CodeSource`]:
/// construction parses immediately and errors on an unknown language or a
/// failed parse, so every method afterwards has a tree.
pub struct CodeAst {
    source: CodeSource<'static>,
}

impl CodeAst {
    /// Parse `source` for `language` (name or alias: `"python"`, `"rust"`,
    /// `"c++"`, …).
    pub fn new(source: impl Into<String>, language: impl Into<String>) -> Result<Self> {
        Self::parse_owned(source.into(), language.into())
    }

    fn parse_owned(source: String, language: String) -> Result<Self> {
        let source = CodeSource::with_language(source, language);
        if source.treesitter_info().is_none() {
            return Err(Error::engine(format!(
                "unknown or non-tree-sitter language: {:?}",
                source.requested_language().unwrap_or_default()
            )));
        }
        if !matches!(source.tree(), ParseOutcome::Parsed(_)) {
            return Err(Error::engine(format!(
                "failed to parse source as {:?}",
                source.requested_language().unwrap_or_default()
            )));
        }
        Ok(Self { source })
    }

    /// The parsed tree (validated at construction).
    fn tree(&self) -> &Tree {
        match self.source.tree() {
            ParseOutcome::Parsed(tree) => tree,
            _ => unreachable!("CodeAst construction validated the parse"),
        }
    }

    /// The language this AST was parsed for.
    pub fn language(&self) -> &str {
        self.source
            .requested_language()
            .expect("CodeAst always has a language")
    }

    /// The source text.
    pub fn source(&self) -> &str {
        self.source.text()
    }

    /// Find every match of a pattern string, compiling it on the spot. Reuses the
    /// AST parse. Prefer [`matches_with`](CodeAst::matches_with) when matching the
    /// same pattern across many ASTs.
    pub fn matches(&self, pattern: &str) -> Result<Vec<CodeMatch>> {
        let cfg = lang::by_name(self.language()).ok_or_else(|| {
            Error::engine(format!(
                "structural matching is not supported for language {:?}",
                self.language()
            ))
        })?;
        let compiled = Pattern::compile(pattern, &cfg)?;
        let raw = compiled.matches_in_tree(self.tree(), self.source());
        Ok(build_sdk_matches(
            self.source(),
            self.source.line_index(),
            raw,
        ))
    }

    /// Find every match of a precompiled [`CodePattern`], reusing both the AST and
    /// the pattern's compilation. This is the fast path for matching the same
    /// pattern against many sources.
    pub fn matches_with(&self, pattern: &CodePattern) -> Result<Vec<CodeMatch>> {
        let same_grammar = self
            .source
            .info()
            .is_some_and(|info| std::ptr::eq(info, pattern.pattern.lang_info()));
        if !same_grammar {
            return Err(Error::engine(format!(
                "CodePattern language {:?} does not match AST language {:?}",
                pattern.language,
                self.language()
            )));
        }
        let raw = pattern.pattern.matches_in_tree(self.tree(), self.source());
        Ok(build_sdk_matches(
            self.source(),
            self.source.line_index(),
            raw,
        ))
    }

    /// Split into syntax-aware chunks, reusing the parse.
    ///
    /// - `chunk_size`: target size in bytes.
    /// - `min_chunk_size`: minimum size in bytes (defaults to `chunk_size / 2`).
    /// - `chunk_overlap`: overlap between consecutive chunks in bytes.
    pub fn split(
        &self,
        chunk_size: usize,
        min_chunk_size: Option<usize>,
        chunk_overlap: Option<usize>,
    ) -> Result<Vec<Chunk>> {
        let chunker = RecursiveChunker::new(RecursiveSplitConfig::default())
            .map_err(|e| Error::engine(format!("failed to create chunker: {e}")))?;
        let config = RecursiveChunkConfig {
            chunk_size,
            min_chunk_size,
            chunk_overlap,
        };
        let raw = chunker.split(&self.source, config);
        Ok(raw
            .into_iter()
            .map(|c| {
                Chunk::new(
                    c.range.start..c.range.end,
                    TextPosition {
                        byte_offset: c.range.start,
                        char_offset: c.start.char_offset,
                        line: c.start.line,
                        column: c.start.column,
                    },
                    TextPosition {
                        byte_offset: c.range.end,
                        char_offset: c.end.char_offset,
                        line: c.end.line,
                        column: c.end.column,
                    },
                )
            })
            .collect())
    }

    /// Extract indexable terms (identifiers + string-literal content, ≥ `min_len`
    /// chars, deduped), reusing the parse. Suitable for building an external
    /// prefilter index (FTS / n-grams).
    pub fn index_terms(&self, min_len: usize) -> Vec<String> {
        index_terms_in_tree(self.tree(), self.source(), min_len)
    }
}

/// The result of [`CodePattern::match_file`]: the parsed [`CodeAst`] and every
/// match found in the file. The file content is `file_match.ast.source()`.
pub struct FileMatch {
    /// The path that was matched.
    pub path: String,
    /// The parsed AST — reuse it to call [`split`](CodeAst::split),
    /// [`matches`](CodeAst::matches), or [`index_terms`](CodeAst::index_terms)
    /// without re-parsing.
    pub ast: CodeAst,
    /// The matches found (always at least one — [`CodePattern::match_file`]
    /// returns `None` when there are none).
    pub matches: Vec<CodeMatch>,
}

// ─── Free functions ───────────────────────────────────────────────────────────

/// One-shot convenience: parse `source` for `language` and return all matches of
/// `pattern`. Equivalent to `CodeAst::new(source, language)?.matches(pattern)`.
///
/// Prefer [`CodePattern`] when matching the same pattern against many sources.
pub fn match_code(pattern: &str, source: &str, language: &str) -> Result<Vec<CodeMatch>> {
    let ast = CodeAst::parse_owned(source.to_string(), language.to_string())?;
    ast.matches(pattern)
}

/// One-shot convenience: parse `source` for `language` and return indexable terms
/// (identifiers + string-literal content, ≥ `min_len` chars, deduped).
///
/// Equivalent to `CodeAst::new(source, language)?.index_terms(min_len)`.
/// Prefer [`CodeAst::index_terms`] when you have an existing parse.
pub fn index_terms(source: &str, language: &str, min_len: usize) -> Result<Vec<String>> {
    let ast = CodeAst::parse_owned(source.to_string(), language.to_string())?;
    Ok(ast.index_terms(min_len))
}
